非同期処理(Javascript)の課題提出

yokomaru opened the pull request:
https://bootcamp.fjord.jp/products/21124/

yokomaru committed:
add: npm node-sqlite3

yokomaru committed:
add: callback
    - 一旦ベタ書きで動く処理を追加

yokomaru committed:
fix: database name

- データベース名を修正
- if (error )を削除

yokomaru committed:
addd: callback presence error

- 一旦エラーありのコールバック処理を追加

yokomaru committed:
fix: eachの部分を修正

yokomaru committed:
add: promise

yokomaru committed:
add: 実行処理を関数にまとめる

yokomaru committed:
fix: DBを実行関数内で定義して各操作に関数に渡す

yokomaru committed:
add: エラー処理も追加

yokomaru committed:
add: callbackにエラー処理も追加

yokomaru committed:
add: async

yokomaru committed:
fix: error async

エラー処理を追加

yokomaru committed:
add: async error handling

- 起こしているエラーのみキャッチする

yokomaru committed:
fix: callback

- クエリを変数にまとめる
- 実行処理を関数にまとめる追加
- エラーありとなしをひとつのファイルにまとめる

yokomaru committed:
fix: promise
- fix function and variable names
- add console message
- add insert params

yokomaru committed:
fix: async
- function and variable name
- display massage

yokomaru committed:
add: error handling

yokomaru committed:
fix: remove Promise Chain Nests

- プロミスチェーン内でネストしていたので修正

yokomaru committed:
add: AUTOINCREMENT

- promiseとasyncには追加してあったが漏れていたため追加

yokomaru committed:
fix: add reject and resolve closeDB

- closeDBでreject と resolve を定義できておらず、ずっとpending状態になっていた

yokomaru committed:
fix: change function name

yokomaru committed:
add: run as node command

- nodeコマンドとして実行できるようにシバンを追加
- ファイルの権限を変更

yokomaru committed:
fix: variable name return promise

- runの実行後にはStatementオブジェクトが返ってくるが、getではDataBaseオブジェクトが返ってくる。
-どちらもどのオブジェクトが返ってくるかは変数として見える必要はないため、返却行という意味合いでRowを使用する

yokomaru committed:
fix: rename execute function name

- Withをつけると長くなりすぎるため、DBOperationの前にSuccessとErrorをつける

yokomaru committed:
fix: display promise error

- エラーオブジェクトではなく、エラーメッセージを表示するように修正

yokomaru committed:
fix: callback error process

- エラーメッセージのみ表示させる
- 正常処理も省略せず記載する

yokomaru committed:
fix: callback error process

- エラーメッセージのみ表示させる
- 正常処理も省略せず記載する

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,108 @@
+import timers from "timers/promises";
+import sqlite3 from "sqlite3";
+
+const createQuery =
+  "CREATE TABLE books (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT UNIQUE)";
+const insertQuery = "INSERT INTO books (title) values ($1)";
+const selectQuery = "SELECT id, title FROM books";
+const errorSelectQuery = "SELECT id, titl FROM books";
+const dropQuery = "DROP TABLE books";
+const insertParam = { $1: "Title1" };
```
他のファイルと定義が重複しています。

yokomaru replied:
**1と2と3の各ファイルを成功失敗で分ける場合**という意味で解釈してしまっていました🙇‍♀️

2と3で共通している定数と関数の定義は以下コミットにて処理を分けて定義しました。

- 関数
[9c4f8fa](https://github.com/yokomaru/js-practices/pull/3/commits/9c4f8fae4b31428ae87ecd74a34a240316f019de)

- 定数
[63e795b](https://github.com/yokomaru/js-practices/pull/3/commits/63e795bc93ff1f26c700766b4ae78fc0411661fe)

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,108 @@
+const dropQuery = "DROP TABLE books";
+const insertParam = { $1: "Title1" };
+
+function runDB(db, query, param) {
```
`runDB` という名前だと database を run するかのように見えてしまいます。`closeDB` はそれで意味的にあっていますが、`runDB` や `getDB` はあっていなさそうです。

yokomaru replied:
`runDB` → `runQuery`

- [Database.run](https://github.com/TryGhost/node-sqlite3/wiki/API#runsql--param---callback)の説明を参考に、SQLクエリを実行するという意味合いにしたかったので`runQuery`に変更いたしました。[1fd521b](https://github.com/yokomaru/js-practices/pull/3/commits/1fd521b790fb39715998416d553a2ae58ea96563)


`getDB` →　`getFirstRow`

- [Database.get](https://github.com/TryGhost/node-sqlite3/wiki/API#getsql--param---callback)を参考に、getは最初の行オブジェクトを取得したいので`getFirstRow`に修正いたしました。
[10c48c7](https://github.com/yokomaru/js-practices/pull/3/commits/10c48c706624c2825b3310dfae17ac8ea919367f)

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,108 @@
+  const db = new sqlite3.Database(":memory:");
+
+  runDB(db, createQuery)
+    .then(() => {
+      return runDB(db, insertQuery, insertParam);
+    })
```
アロー関数の中身が return しかないときは簡潔文体を使ってください。

yokomaru replied:
以下で対応いたしました。

[af34f7c](https://github.com/yokomaru/js-practices/pull/3/commits/af34f7cca8fea6199263f6f63855074d37c77df2)

[00a1417](https://github.com/yokomaru/js-practices/pull/3/commits/00a14178f7c82a8d75cde89971243171745a571d)

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,108 @@
+      return runDB(db, insertQuery, insertParam);
+    })
+    .then((result) => {
+      console.log(`this.lastID: ${result.lastID}`);
```
`this.lastID` と出力されていますがそうではなさそうです。

yokomaru replied:
おっしゃる通りpromiseとasyncの場合thisの値ではなく、Promiseの実行後に渡されたStatementオブジェクトの値でした。

console.logの表記`this.LastID`を`Statement.lastID`にするか、変数名に合わせて`result.lastID`に修正するか悩みましたが、ここではlastIDということがわかればいいかな？と思ったので`lastID: `に修正いたしました。[a414acf](https://github.com/yokomaru/js-practices/pull/3/commits/a414acfabbffba7945a94ac284ff1d21349c3cf6)

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,92 @@
+const dropQuery = "DROP TABLE books";
+const insertParam = { $1: "Title1" };
+
+const runDB = (db, query, param) => {
+  return new Promise((resolve, reject) => {
+    db.run(query, param, function (error) {
+      if (error) {
+        reject(error);
+      } else {
+        resolve(this);
+      }
+    });
+  });
+};
+
+const getDB = (db, query) => {
+  return new Promise((resolve, reject) => {
+    db.get(query, (error, row) => {
+      if (error) {
+        reject(error);
+      } else {
+        resolve(row);
+      }
+    });
+  });
+};
+
+function closeDB(db) {
+  return new Promise((resolve, reject) => {
+    db.close((error) => {
+      if (error) {
+        reject(error);
+      } else {
+        resolve();
+      }
+    });
+  });
+}
```
他のファイルと定義が重複しています。課題の注意点にもコピーしないように書かれていますよ。https://bootcamp.fjord.jp/pages/511#%E6%B3%A8%E6%84%8F%E7%82%B9-3

また、`runDB` と `getDB` はアロー関数と変数宣言で定義されているのに、`closeDB` は function 文で定義されていて一貫性がないです。

yokomaru replied:
>他のファイルと定義が重複しています。課題の注意点にもコピーしないように書かれていますよ。https://bootcamp.fjord.jp/pages/511#%E6%B3%A8%E6%84%8F%E7%82%B9-3

以下で対応いたしました。

https://github.com/yokomaru/js-practices/pull/3#discussion_r1980902764

> また、runDB と getDB はアロー関数と変数宣言で定義されているのに、closeDB は function 文で定義されていて一貫性がないです。

見落としていました、すみません🙇
functionで定義しているものは関数はアロー関数に修正いたしました。(別ファイルに切り出した際に修正しています)

[9c4f8fa](https://github.com/yokomaru/js-practices/pull/3/commits/9c4f8fae4b31428ae87ecd74a34a240316f019de)

yokomaru replied:
すみません、promise.js内に修正漏れがあったので追加で修正しております。

[73842c4](https://github.com/yokomaru/js-practices/pull/3/commits/73842c4d57a8e2de3c4c7e939bed8867c324dbb6)

また、callback.jsの18行目とdb_operation.jsの`runQuery`のコールバックの箇所はfunction式でないと`this.lastID`が取得できないとリファレンスに記載があったので残しております。
https://github.com/TryGhost/node-sqlite3/wiki/API#runsql--param---callback

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,92 @@
+  const db = new sqlite3.Database(":memory:");
+
+  await runDB(db, createQuery);
+  const insertedRow = await runDB(db, insertQuery, insertParam);
```
Promise のプログラムと結果を受ける変数の名前が異なっています。また、`runDB` の実行結果として受け取れるのは row ではないので名前も適当ではなさそうです。

yokomaru replied:
`runDB`で帰ってくるのはStatementオブジェクトですが、`insertedStatement`だとSQLのインサート文そのもののように見えると思ったため、insertedResultに修正いたしました。[5ebc3bd](https://github.com/yokomaru/js-practices/pull/3/commits/5ebc3bd34214c5b27d2f91b139b4efc74b9f231d)

また、`getDB`の結果は`selectedRow`としていましたが、正しくは一番最初の行を取得したかったので`firstRowに`修正いたしました。[b0464e5](https://github.com/yokomaru/js-practices/pull/3/commits/b0464e52e0a2512d69037f47631cab07b42f3292)

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,92 @@
+  try {
+    const insertedRow = await runDB(db, insertQuery, insertParam);
+    console.log(`this.lastID: ${insertedRow.lastID}`);
+  } catch (error) {
+    if (error.code == "SQLITE_CONSTRAINT") {
+      console.error(error.message);
+    }
+  }
```
例外が握り潰されています。例外を握り潰してはいけません。例外処理を行っている意味がなくなってしまいます。捕捉したい例外のみを捕捉するようにしてください。

yokomaru replied:
- MDNの[特定のエラーを処理する](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Error#%E7%89%B9%E5%AE%9A%E3%81%AE%E3%82%A8%E3%83%A9%E3%83%BC%E3%82%92%E5%87%A6%E7%90%86%E3%81%99%E3%82%8B)を参考に以下を対応いたしました。
     - エラークラスのインスタンスかどうか かつ エラーコードがSQLITE_CONSTRAINTかどうかを分岐する
    - 当てはまらない場合はエラーを再スローする

[1cb907c](https://github.com/yokomaru/js-practices/pull/3/commits/1cb907c1b2b97814d90860da583ec5eaf9babded)

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,92 @@
+    const insertedRow = await runDB(db, insertQuery, insertParam);
+    console.log(`this.lastID: ${insertedRow.lastID}`);
+  } catch (error) {
+    if (error.code == "SQLITE_CONSTRAINT") {
```
`error` にプロパティアクセスできないような値が格納されていると `error.code` の部分でエラーが起きてしまいます。

yokomaru replied:
[オプショナルチェーン ](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Operators/Optional_chaining)を使用して、`error.?code`とすることでエラープロパティアクセスエラーが起こらないように修正いたしました [1cb907c](https://github.com/yokomaru/js-practices/pull/3/commits/1cb907c1b2b97814d90860da583ec5eaf9babded)

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,92 @@
+
+console.log("Success");
+executeSuccessDBOperation();
+await timers.setTimeout(100);
```
async/await のプログラムでは `setTimeout` の呼び出しは必要ないはずです。なぜかを考えて理由をコメントしてください。

yokomaru replied:
> async/await のプログラムでは setTimeout の呼び出しは必要ないはずです。なぜかを考えて理由をコメントしてください。

async/awaitを使用する場合、awaitがPromiseの結果（resolveまたはreject）が返却されるまで処理を待機するため、setTimeout (正確にはPromise)を呼び出して一時的に処理を待機させる必要がないという認識ですが、合っていますでしょうか？

こちら、実装した当時はあまり理解できておらず、`executeSuccessDBOperation();`の後のconsole.logを表示するためにSetTimeOutを使ってしまってました。

```js
executeSuccessDBOperation();
await timers.setTimeout(100);
console.log("-------------------------------------");
console.log("Error");
executeErrorDBOperation();
```
async関数である`executeSuccessDBOperation ()`自体もPromiseを返却するため、`executeSuccessDBOperation()`と`executeErrorDBOperation()`にもawaitを付与してPromiseが返却するまで待機させた後、console.logを表示するように修正いたしました。

[73fdac8](https://github.com/yokomaru/js-practices/pull/3/commits/73fdac8fc025ffb5b6badb404352e0ffb1baec63)

yokomaru committed:
add: memo to fix

- 修正箇所を一旦メモして後ほど消す

yokomaru committed:
fix: cut query constants into separate files

- クエリ定数を別ファイルに切りだす

yokomaru committed:
fix: cut out DB operation to a separate file

- DB操作関数を別ファイルに切り出す

yokomaru committed:
fix: variable names to receive results of DB operation queries

- RowではなくStateMentオブジェクトトが帰ってくる
- ステートメントを実行した結果を保持する変数なのでResultを使用する

yokomaru committed:
fix: function name

- sqlite3の関数runはステートメントを実行するのでrunStatement
- getはクエリ実行後最初の結果行を取得するのでgetFirstRow

yokomaru committed:
fix: async error

- 例外が握り潰されていた部分を細くしたいエラーのみキャッチし、それ以外は再度エラーをスローする
- ?. で.codeにアクセスできない時にエラーにしないようにする

yokomaru committed:
fix: add await to asyc functions and delete setTimeout

- executeSuccessDBOperationとexecuteErrorDBOperationが完了するまでconsole.logの表示を止めたいので、awaitでPromiseが返却するまで待機する

yokomaru committed:
fix: this.lastID

db.runで帰ってくるthisはStatementオブジェクトなので、メッセージもStatement.lastIDとして表示する(callbackとpromiseと統一する)

yokomaru committed:
delete : import timers

asyncでtimersを使用しなくなったので削除

yokomaru committed:
fix: convert Arrow functions to concise syntax

- return のみのアロー関数は簡潔構文に修正する

yokomaru committed:
fix: style and unnecessary comments

- eslintで指摘された箇所を修正
- 追加した修正メモを削除

yokomaru committed:
fix: function name runStatement

SQLクエリを実行するという意味に変更するため、runQueryにする

yokomaru committed:
fix:
variable name insertResult

- インサート実行後の結果を保持するためinsertedResultに変更

yokomaru committed:
fix: variable name selectResult

getFirstRowの結果を取得するので変数名をFirstRowに変更

yokomaru committed:
fix: console log lastID

- this.lastID、Statement.lastIDとしていたがlastIDであることがわかれば良いのでlastIDに修正

yokomaru committed:
fix: function expression to arrow expression in promise.js

- executeSuccessDBOperationとexecuteErrorDBOperationをアロー式に修正

yokomaru committed:
fix:  convert Arrow functions to concise syntax in callback.js

- db.close()のアロー構文も簡潔構文に修正

yokomaru committed:
fix: converting if statements to conditional operator

- コードをより簡潔にするためにif文を三項演算子に修正

cafedomancer commented on 03.asynchronous/query.js:
```diff
@@ -0,0 +1,7 @@
+export const createQuery =
+  "CREATE TABLE books (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT UNIQUE)";
+export const insertQuery = "INSERT INTO books (title) values ($1)";
```
コメントし忘れていましたが、SQL のキーワード部分は大文字で書いてください。

```suggestion
export const insertQuery = "INSERT INTO books (title) VALUES ($1)";
```

yokomaru replied:
こちら見落としておりました💦ご指摘ありがとうございます！
以下で対応いたしました。

[7ad4283](https://github.com/yokomaru/js-practices/pull/3/commits/7ad4283d62051a6a1059f9d5551f67160bc95a91)

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,55 @@
+    {
+      db.run(insertQuery, insertParam, function () {
+        console.log(`lastID: ${this.lastID}`);
+        db.get(selectQuery, (error, result) => {
```
使われていない変数 (`error`) はそれが分かるような命名規則を適用してください。`result` についてはもっと具体的な名前をつけたほうがよいと思います。

yokomaru replied:
調べまして、引数として設定するが関数内では使われない場合は`_`を使用することを理解いたしました。
以下で修正しております。

[07b0013](https://github.com/yokomaru/js-practices/pull/3/commits/07b00137f0e1865e0194f5449f1c868dd0921788)

また、`result`につきましては以下で指摘いただいた`firstRow`と合わせて`selectQuery `の結果を受け取る変数はUS全て`row`に変更いたしました。
https://github.com/yokomaru/js-practices/pull/3#discussion_r1985758182

以下で対応しております。
[67dc6e2](https://github.com/yokomaru/js-practices/pull/3/commits/67dc6e221fd7d3ee1e28d951268b50207a4789bd)

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,55 @@
+        console.log(`lastID: ${this.lastID}`);
+        db.get(selectQuery, (error, result) => {
+          console.log(`${result.id}: ${result.title}`);
+          db.run(dropQuery, () => db.close());
```
戻り値のないアロー関数はそのことを明示するためにブロック文体で書いてください。

yokomaru replied:
アロー関数にて、ブロックなしは`return`が省略されているということを理解いたしました。
いくつか戻り値なしだがブロックなしになっていたものがあったので以下で合わせて修正いたしました。

[665f8cf](https://github.com/yokomaru/js-practices/pull/3/commits/665f8cfd041e2d48a3b5604440fa48a1aa49901b)

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,55 @@
+    {
+      db.run(insertQuery, insertParam, () => {
+        db.run(insertQuery, insertParam, (error, result) => {
+          error
+            ? console.error(error.message)
+            : console.log(`lastID: ${result.lastID}`);
```
式全体の評価結果が必要ないのなら条件演算子ではなく if 文で書いてください。条件演算子はその式全体の評価結果が必要な場合に使われることが多いです。

yokomaru replied:
条件演算子が式全体の評価結果が必要な場合に使用する旨と条件に応じて異なるコードを実行したい場合はif文を使う必要がある旨理解いたしました。
見直して、該当するコードを以下で対応いたしました。
[fd3ab92](https://github.com/yokomaru/js-practices/pull/3/commits/fd3ab92f9c79c38543829305fcdccb654b03fafc)

cafedomancer commented on 03.asynchronous/db_operation.js:
```diff
@@ -0,0 +1,19 @@
+  });
+};
+
+export const getFirstRow = (db, query) => {
```
元の get 関数ではパラメータも扱えるようになっているはずです。

yokomaru replied:
APIリファレンスを見直し、パラメーターも扱えるようになっている旨確認いたしました。
以下でparamを引数として追加しております。

[8a1f271](https://github.com/yokomaru/js-practices/pull/3/commits/8a1f2710fdbcea68dbde9d540e18da162e03f963)

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,51 @@
+
+  runQuery(db, createQuery)
+    .then(() => runQuery(db, insertQuery, insertParam))
+    .then((insertedResult) => {
```
inserted でない result があるなら inserted をつけたほうがよいと思いますが、ないのでいらないと思います。

yokomaru replied:
>inserted でない result があるなら inserted をつけたほうがよいと思いますが、ないのでいらない

変数の付け方にとても悩んでいたので、考え方を共有いただきとてもありがたいです！
inserted以外のresultはないので、単純なクエリ実行結果の`result`に修正いたしました。

[edb3678](https://github.com/yokomaru/js-practices/pull/3/commits/edb36783ae842233f57ed344f4403a41c357b150)

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,51 @@
+      console.log(`lastID: ${insertedResult.lastID}`);
+      return getFirstRow(db, selectQuery);
+    })
+    .then((firstRow) => {
```
get 関数に対応する関数に first が含まれている (`getFirstRow`) のは挙動を表現するものとしてよい気がするのですが、ここの `firstRow` は別に first でない row があることを期待しているわけではないので、first はつけないほうがよい気がしました。first 以外の row もあるかのように見えてしまいます。

yokomaru replied:
以下で対応しております。
[67dc6e2](https://github.com/yokomaru/js-practices/pull/3/commits/67dc6e221fd7d3ee1e28d951268b50207a4789bd)

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,57 @@
+    const insertedResult = await runQuery(db, insertQuery, insertParam);
+    console.log(`lastID: ${insertedResult.lastID}`);
+  } catch (error) {
+    if (error instanceof Error && error?.code == "SQLITE_CONSTRAINT") {
```
`&&` の左辺が true になる時点で `error` に Error オブジェクトが格納されていることが決まるので、右辺の `error?.code` のオプショナルチェーンは無駄です。

yokomaru replied:
こちらおっしゃる通りでした。
`.code`プロパティを持っていないErrorオブジェクトが来た場合もプロパティアクセスエラーになると勘違いしていたためオプショナルチェーンが必要だと思っていたのですが、その場合は`undifined`になりfalse判定されるので問題ありませんでした。

以下で対応いたしました。

[4e5ad85](https://github.com/yokomaru/js-practices/pull/3/commits/4e5ad8514a60c3e22405a25ac7ad515fb86a6519)

yokomaru committed:
fix: write SQL keyword parts in uppercase

- insertのvaluesが小文字になっていたが、SQL のキーワード部分は大文字で書くので修正

yokomaru committed:
fix: change all variables in the result of selectQuery to row

- getFirstRow(selectQuery)の実行結果を受け取る変数は全てrowに変更
    - resultだと抽象的すぎる、かつfirstRowだとfirstじゃないrowがあるとfirst以外のrowがあるように見えるため、rowに統一

yokomaru committed:
fix: add "param" parameter to getResult

- 元の get 関数ではパラメータも扱えるようになっているため、paramパラメーターを追加

yokomaru committed:
fix: change unused variable “error” to “_”.

- 未使用の変数errorはアンダーバー(_)にする

yokomaru committed:
fix: change the Arrow function with no return value to a block statement style

- 戻り値がないことを明示するためにアロー関数をブロック文体で書く

yokomaru committed:
fix: rewrite conditional operators into if statements

- 評価結果ではなく実行処理を分岐したいため?演算子ではなくif文に書き直す

yokomaru committed:
fix: variable "insertedResult" to "result"

- inserted でない result があるように見えるので、resultに修正

yokomaru committed:
fix:delete an optional chain

- 'error instanceof Error'でエラーオブジェクトが帰ってくるのがわかるので、codeがnullまたはundefinedの可能性はないのオプショナルチェーンを削除する

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,63 @@
+  db.run(createQuery, () => {
+    {
+      db.run(insertQuery, insertParam, () => {
+        db.run(insertQuery, insertParam, (error, result) => {
```
(コメントし忘れていたのかもしれませんが) run の実行結果はコールバックの引数 (`result`) では取れないですよ。

yokomaru replied:
こちら、error処理の中の正常処理の確認が不足しておりました💦
APIの仕様を再度読み直し、runの実行結果はコールバックの引数としてではなく、functionを使用しthis(Statementオブジェクト)を参照する必要があることを理解いたしました。

[763b8dc](https://github.com/yokomaru/js-practices/pull/3/commits/763b8dc0ab6e5f4399b1739f5af259a929c532bc)

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,67 @@
+import { runQuery, getFirstRow, closeDB } from "./db_operation.js";
+
+const executeSuccessDBOperation = () => {
+  const db = new sqlite3.Database(":memory:");
+
+  runQuery(db, createQuery)
```
コールバックのプログラムではこの処理間に空行がなかったのに、Promsie のプログラムでは空行があります。どちらでもいいですが、スタイルに一貫性を持たせてください。

yokomaru replied:
async.js、callback.jsに合わせて空行を削除いたしました。

[79ece4c](https://github.com/yokomaru/js-practices/pull/3/commits/79ece4c41b07c61be4bf1809fed6e8d2916d69c5)

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,67 @@
+      console.log(`${row.id}: ${row.title}`);
+      return runQuery(db, dropQuery);
+    })
+    .then(() => {
+      closeDB(db);
+    });
```
後続の処理を追加することも考えて `closeDB()` が返す Promise の完了も待つようにしてください。

yokomaru replied:
ご指摘ありがとうございます。
こちら、前回の修正の際、次の処理に引数を渡す必要がない(戻り値不要)のためreturnを削除してしまったのですが、 `then` 内で `return` をしないと生成されたPromiseの完了を待たずに次の処理にうつってしまうことを改めて理解いたしました。
対応としては、`{}` を外しPromiseオブジェクトを `return` させるように修正いたしました。

[a2ad151](https://github.com/yokomaru/js-practices/pull/3/commits/a2ad151b7e352f0989964c11229ab543466592cc)

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,67 @@
+  const db = new sqlite3.Database(":memory:");
+
+  runQuery(db, createQuery)
+    .then(() => {
+      runQuery(db, insertQuery, insertParam);
+    })
```
この書き方だと `runQuery()` の返す Promise の完了を待たずに後続の処理が実行される可能性があります。

yokomaru replied:
https://github.com/yokomaru/js-practices/pull/3#discussion_r1996861095 と同様に `{}` を外し `return` させるように修正いたしました。(同じような箇所も合わせて修正しております)

[e0f7ef6](https://github.com/yokomaru/js-practices/pull/3/commits/e0f7ef68d26c977490e67b4092672781e2797069)

yokomaru committed:
fix: modify an Arrow expression to a function expression

- db.runの実行結果は引数のresultで受け取れず、db.runの中で、functionを使用する必要があるため修正
    - https://github.com/TryGhost/node-sqlite3/wiki/API#runsql--param---callback
    - ステートメントが実行された後にコールバックが実行され、呼び出し元(実行コンテキスト)がステートメントオブジェクトとなる
    -  アロー関数はthisの値が宣言時で確定するため、上記の処理を経てthisを受け取るためにはfunction形式にする必要がある

yokomaru committed:
fix: wait for completion of Promise returned by closeDB()

- 後続の処理を追加することも考えて closeDBの処理をreturnする
- {}なしのアロー関数はreturn と同義のため{}を削除

yokomaru committed:
fix: unify the style of line breaks after DB definitions

- async.js、callback.jsではconst db後は空行なく、統一感がないため空行を削除する

yokomaru committed:
fix:  wait for the Promise returned by runQuery() to complete before executing the subsequent process

- 次の処理に引数を渡す必要がない(戻り値不要)のためreturnを削除してしまったが、 then内でreturnをしないと生成されたPromiseの完了を待たずに次の処理にうつってしまう
- {}を外し、returnさせるように修正
